src/widget/scroll.rs

```diff
@@ -56,13 +56,21 @@ impl ScrollDirection {
 struct ScrollBarsState {
     opacity: f64,
     timer_id: TimerToken,
+    hovered: bool,
+    vertical_held: bool,
+    horizontal_held: bool,
+    held_offset: f64,
```

Comment:
The thing that jumps out at me in the PR are these four items.

I have a fairly abstract concern/curiosity: is this the best way to encode the state that we're trying to encode? I might be mistaken, but it seems like a bunch of these states are mutually exclusive. That is, it seems like you are _either_ `hovered` _or_ `vertical_held` _or_ `horizontal_held`. And it seems like `held_offset` refers to an offset in the `y` direction *if* `horizontal_held` is `true`, and an offset in the `x` direction _if_ `vertical_held` is `true`; and it's an invariant that only one of these can be true at a time. (I think; I could be wrong!)

Anyway, given these assumptions, I could imagine replacing these four variables with an enum, that might look something like:

```rust
/// The state of the mouse with respect to the scrollbars
enum Mouse {
    /// The mouse is not interacting with a scrollbar
    None,
    /// the mouse is over the vertical scrollbar
    HoverVertical,
    /// the mouse is over the horizontal scrollbar
    HoverHorizontal,
    /// The mouse is dragging the vertical scrollbar. The `f64` field
    /// is the y-offset where the gesture began.
    DownVertical(f64),
    /// The mouse is dragging the horizontal scrollbar. The `f64` field
    /// is the x-offset where the gesture began.
    DownHorizontal(f64),
}
```

With something like this, you could have a single value that represented the current state, and your handling of events could just be a matter of transitioning between states and acting accordingly. As a bonus, this makes illegal states impossible to create; you cannot have `held_vertical` and `held_horizontal` both be true, `held_offset` can't have a value unless something is held, etcetera.

You can also add some simple methods to this enum, like

```rust
impl Mouse {
    fn is_held(self) -> bool {
        match self {
            Mouse::DownVertical(_) | Mouse::DownHorizontal(_) => true,
            _ => false,
        }
    }
}
```

And so it can still act like a `bool` where that makes sense.


Anyway: this is just a sketch, and it may not correctly model the state that you've written, but I suspect that there is _some_ model like this that will be more robust, more maintainable, and more ergonomic than juggling a bunch of individual values.

Reply:
I *completely* forgot about Rust's magical enums. I frequently wish I had sum types in the languages I use on the daily but I thoroughly forgot about them once I was in a language which *does* have them! I'll look into this right away!

Reply:
I ended up using separate enums for held and hover statuses

Reply:
I ended up using separate enums for hovered and held statuses